File: ProjectCS/UI/GroupTablePage.cs
using System;
using System.Collections.Generic;
using System.Linq;


namespace SessionResultsCS.UI
{
    public class GroupTablePage : IPage
    {
        private Faculty _faculty;

        public GroupTablePage(Faculty faculty)
        {
            _faculty = faculty;
        }

        public void Start()
        {
            if (_faculty.GetSpecialities().All(speciality => speciality.GetGroupCount() == 0))
            {
                Console.WriteLine("В факультете отсутствуют группы");
                return;
            }

            GroupTable groupTable = CreateGroupTable();

            PrintTable(groupTable);

            Console.WriteLine("Действия: ");
            new MenuActionPicker([
                ("Экспорт отчёта в TXT файл", () => { ExportToTXT(groupTable); }),
                ("Назад в главное меню", () => {}),
            ]).PickAndRun();
        }


        private GroupTable CreateGroupTable()
        {
            Console.WriteLine("Выберите специальность: ");
            ItemPicker specialityPicker = new ItemPicker(_faculty.GetSpecialities().Select(x => x.ToString()));
            Speciality speciality = _faculty.GetSpecialityAt(specialityPicker.Pick());


            Console.WriteLine("Выберите группу: ");
            ItemPicker groupPicker = new ItemPicker(speciality.GetGroups().Select(x => x.ToString()));
            Group group = speciality.GetGroupAt(groupPicker.Pick());


            Console.WriteLine("Выберите дисциплины: ");
            foreach (var (index, discipline) in group.GetDisciplineList().Index())
            {
                Console.WriteLine($"{index + 1}. {discipline}");
            }
            SortedSet<int> indices = new SortedSet<int>();
            bool readIndices = false;
            while (!readIndices)
            {
                indices.Clear();
                readIndices = true;

                Console.Write("Введите номера дисциплин через запятую (например: 1,3,4): ");
                string input = Console.ReadLine()!;
                foreach (string el in input.Split(","))
                {
                    if (int.TryParse(el, out int value)
                        && (value >= 1 && value <= group.GetDisciplineList().GetSize())
                        && !indices.Contains(value - 1))
                    {
                        indices.Add(value - 1);
                    }
                    else
                    {
                        readIndices = false;
                        Console.WriteLine("Номера введены неверно");
                        break;
                    }
                }
            }
            DisciplineList disciplines = new DisciplineList(indices.Select(index => group.GetDisciplineList().GetDisciplineAt(index)));


            ItemPicker sortPicker = new ItemPicker(["Средний балл", .. disciplines.Select(x => x.ToString())]);
            Console.WriteLine("Выберите столбец для сортировки: ");
            int sortSelection = sortPicker.Pick();

            QuickItemPicker sortOrderPicker = new QuickItemPicker(["По возрастанию", "По убыванию"]);
            Console.WriteLine("Выберите порядок сортировки: ");
            SortOrder sortOrder = sortOrderPicker.Pick() switch
            {
                0 => SortOrder.Ascending,
                1 => SortOrder.Descending
            };

            QuickItemPicker includeOnlyFailingPicker = new QuickItemPicker(["Включить всех студентов", "Включить только неуспевающих"]);
            Console.WriteLine("Выберите, включать ли всех студентов: ");
            bool includeOnlyFailing = includeOnlyFailingPicker.Pick() == 1;


            GroupTable groupTable = new GroupTable(group, includeOnlyFailing);
            groupTable.Disciplines = disciplines;

            if (sortSelection == 0)
            {
                groupTable.SortByAverage(sortOrder);
            }
            else
            {
                groupTable.SortByDiscipline(disciplines.GetDisciplineAt(sortSelection - 1), sortOrder);
            }

            return groupTable;
        }

        private void PrintTable(GroupTable groupTable)
        {
            GroupTableData data = groupTable.GetTableData();
            Console.WriteLine(StyledTextTableRenderer.Default.RenderTable(data));
            Console.WriteLine();

            for (int i = 0; i < data.Disciplines.Length; i++)
            {
                Console.WriteLine($"{i + 1} - {data.Disciplines[i].Name}");
            }
        }

        private void ExportToTXT(GroupTable groupTable)
        {
            var (fileName, title, body, includeDate) = GetReportExportData(".txt");

            QuickItemPicker tableStylePicker = new QuickItemPicker(["Обычный", "С двойным контуром"]);
            Console.WriteLine("Выберите стиль таблицы: ");
            int selectedTableStyle = tableStylePicker.Pick();

            Console.Write("Введите симбол таблицы на пересечении (по умолчанию: +): ");
            char crossSymbol = Console.ReadKey().KeyChar;
            if (crossSymbol == '\r') crossSymbol = '+';
            Console.WriteLine();

            Console.Write("Введите горизонтальный симбол таблицы (по умолчанию: -): ");
            char horizontalSymbol = Console.ReadKey().KeyChar;
            if (horizontalSymbol == '\r') horizontalSymbol = '-';
            Console.WriteLine();

            Console.Write("Введите вертикальный симбол таблицы (по умолчанию: |): ");
            char verticalSymbol = Console.ReadKey().KeyChar;
            if (verticalSymbol == '\r') verticalSymbol = '|';
            Console.WriteLine();

            ITextTableRenderer renderer;
            if (selectedTableStyle == 0)
            {
                renderer = new StyledTextTableRenderer(crossSymbol, horizontalSymbol, verticalSymbol);
            }
            else
            {
                renderer = new OutlinedTextTableRenderer(crossSymbol, horizontalSymbol, verticalSymbol);
            }

            IReportExporter exporter = new TXTReportExporter(fileName, renderer, title, body, includeDate);
            exporter.Export(groupTable);
            Console.WriteLine($"Отчёт сохранён в {fileName}");
        }

        private (string FileName, string Title, string Body, bool IncludeDate) GetReportExportData(string fileExtenstion)
        {
            Console.Write("Введите имя файла для сохранения отчёта: ");
            string fileName = Console.ReadLine()!;
            if (!fileName.EndsWith(fileExtenstion))
            {
                fileName += fileExtenstion;
            }

            Console.Write("Введите заголовок отчёта или пустую строку: ");
            string title = Console.ReadLine()!;

            Console.Write("Введите текст отчёта или пустую строку: ");
            string body = Console.ReadLine()!;

            QuickItemPicker includeDatePicker = new QuickItemPicker(["Не включать дату в отчёт", "Включать дату в отчёт"]);
            Console.WriteLine("Выберите, включать ли текущую дату в отчёт: ");
            bool includeDate = includeDatePicker.Pick() == 1;

            return (fileName, title, body, includeDate);
        }
    }
}

File: Project/UI/GroupTablePage.cpp
#include "UI/GroupTablePage.hpp"

#include "GroupTable/ITextTableRenderer.hpp"
#include "GroupTable/StyledTextTableRenderer.hpp"
#include "GroupTable/OutlinedTextTableRenderer.hpp"
#include "GroupTable/TXTReportExporter.hpp"
#include "UI/MenuActionPicker.hpp"
#include "UI/ItemPicker.hpp"
#include "UI/QuickItemPicker.hpp"

#include <conio.h>
#include <format>
#include <iostream>
#include <set>
#include <sstream>
#include <algorithm>


GroupTablePage::GroupTablePage(std::shared_ptr<Faculty> faculty) {
	m_faculty = faculty;
}

void GroupTablePage::Start() {
	if (std::ranges::all_of(m_faculty->GetSpecialities(),
		[](auto& speciality) { return speciality->GetGroupCount() == 0; })) {
		std::cout << "В факультете отсутствуют группы\n";
		return;
	}

	GroupTable groupTable = CreateGroupTable();

	PrintTable(groupTable);

	std::cout << "Действия:\n";
	MenuActionPicker({
		{ "Экспорт отчёта в TXT файл", [&]() { ExportToTXT(groupTable); } },
		{ "Назад в главное меню", []() {} }
	}).PickAndRun();
}

GroupTable GroupTablePage::CreateGroupTable() {
	std::vector<std::string> specialityStrings;
	for (auto& speciality: m_faculty->GetSpecialities()) {
		if (speciality->GetGroupCount() != 0) {
			specialityStrings.push_back(speciality->ToString());
		}
	}
	ItemPicker specialityPicker(specialityStrings);

	std::cout << "Выберите специальность:\n";
	std::shared_ptr<Speciality> speciality = m_faculty->GetSpecialityAt(specialityPicker.Pick());


	std::vector<std::string> groupStrings;
	for (auto& group: speciality->GetGroups()) {
		groupStrings.push_back(group->ToString());
	}
	ItemPicker groupPicker(groupStrings);

	std::cout << "Выберите группу:\n";
	std::shared_ptr<Group> group = speciality->GetGroupAt(groupPicker.Pick());


	std::cout << "Выберите дисциплины:\n";
	for (int i = 0; i < group->GetDisciplineList().GetSize(); i++) {
		std::cout << std::format("{}. {}\n", i + 1, group->GetDisciplineList().GetDisciplineAt(i)->ToString());
	}

	std::set<int> indices;
	bool readIndices = false;
	while (!readIndices) {
		indices.clear();
		readIndices = true;

		std::cout << "Введите номера дисциплин через запятую (например: 1,3,4): ";
		std::string input;
		std::getline(std::cin, input);

		std::stringstream ss(input);
		std::string nextNumber;
		while (std::getline(ss, nextNumber, ',')) {
			try {
				int value = std::stoi(nextNumber);
				if (value >= 1 && value <= group->GetDisciplineList().GetSize() && !indices.contains(value - 1)) {
					indices.insert(value - 1);
				} else {
					readIndices = false;
					std::cout << "Номера введены неверно\n";
					break;
				}
			} catch (std::invalid_argument) {
				readIndices = false;
				std::cout << "Номера введены неверно\n";
				break;
			} catch (std::out_of_range) {
				readIndices = false;
				std::cout << "Номера введены неверно\n";
				break;
			}
		}
	}

	DisciplineList disciplines;
	for (int index : indices) {
		disciplines.AddDiscipline(group->GetDisciplineList().GetDisciplineAt(index));
	}

	std::vector<std::string> sortOptions = { "Средний балл" };
	for (int i = 0; i < disciplines.GetSize(); i++) {
		sortOptions.push_back(disciplines.GetDisciplineAt(i)->ToString());
	}
	ItemPicker sortPicker(sortOptions);
	std::cout << "Выберите столбец для сортировки:\n";
	int sortSelection = sortPicker.Pick();

	QuickItemPicker sortOrderPicker({"По возрастанию", "По убыванию"});
	std::cout << "Выберите порядок сортировки:\n";
	GroupTable::SortOrder sortOrder = (sortOrderPicker.Pick() == 0) ? GroupTable::SortOrder::eAscending : GroupTable::SortOrder::eDescending;

	QuickItemPicker includeOnlyFailingPicker({"Включить всех студентов", "Включить только неуспевающих"});
	std::cout << "Выберите, включать ли всех студентов:\n";
	bool includeOnlyFailing = (includeOnlyFailingPicker.Pick() == 1);

	GroupTable groupTable(group, includeOnlyFailing);
	groupTable.SetDisciplineList(disciplines);

	if (sortSelection == 0) {
		groupTable.SortByAverage(sortOrder);
	} else {
		groupTable.SortByDiscipline(disciplines.GetDisciplineAt(sortSelection - 1), sortOrder);
	}

	return groupTable;
}

void GroupTablePage::PrintTable(const GroupTable& groupTable) {
	GroupTableData data = groupTable.GetTableData();

	std::cout << StyledTextTableRenderer::GetDefault()->RenderTable(data) << std::endl;
	std::cout << std::endl;

	for (int i = 0; i < data.GetDisciplines().size(); i++) {
		std::cout << std::format("{} - {}\n", i + 1, data.GetDisciplines()[i]->GetName());
	}
}

void GroupTablePage::ExportToTXT(const GroupTable& groupTable) {
	auto [fileName, title, body, includeDate] = GetReportExportData(".txt");

	QuickItemPicker tableStylePicker({"Обычный", "С двойным контуром"});
	std::cout << "Выберите стиль таблицы:\n";
	int selectedTableStyle = tableStylePicker.Pick();

	std::cout << "Введите символ таблицы на пересечении (по умолчанию: +): ";
	char crossSymbol = _getch();
	if (crossSymbol == '\r') crossSymbol = '+';
	std::cout << std::endl;

	std::cout << "Введите горизонтальный символ таблицы (по умолчанию: -): ";
	char horizontalSymbol = _getch();
	if (horizontalSymbol == '\r') horizontalSymbol = '-';
	std::cout << std::endl;

	std::cout << "Введите вертикальный символ таблицы (по умолчанию: |): ";
	char verticalSymbol = _getch();
	if (verticalSymbol == '\r') verticalSymbol = '|';
	std::cout << std::endl;

	std::unique_ptr<ITextTableRenderer> renderer;
	if (selectedTableStyle == 0) {
		renderer = std::make_unique<StyledTextTableRenderer>(crossSymbol, horizontalSymbol, verticalSymbol);
	} else {
		renderer = std::make_unique<OutlinedTextTableRenderer>(crossSymbol, horizontalSymbol, verticalSymbol);
	}

	std::unique_ptr<IReportExporter> exporter = std::make_unique<TXTReportExporter>(fileName, std::move(renderer), title, body, includeDate);
	exporter->Export(groupTable);
	std::cout << std::format("Отчёт сохранён в {}\n", fileName);
}

std::tuple<std::string, std::string, std::string, bool> GroupTablePage::GetReportExportData(const std::string& fileExtension) {
	std::cout << "Введите имя файла для сохранения отчёта: ";
	std::string fileName;
	std::getline(std::cin, fileName);
	if (!fileName.ends_with(fileExtension)) {
		fileName += fileExtension;
	}

	std::cout << "Введите заголовок отчёта или пустую строку: ";
	std::string title;
	std::getline(std::cin, title);

	std::cout << "Введите текст отчёта или пустую строку: ";
	std::string body;
	std::getline(std::cin, body);

	QuickItemPicker includeDatePicker({"Не включать дату в отчёт", "Включать дату в отчёт"});
	std::cout << "Выберите, включать ли текущую дату в отчёт:\n";
	bool includeDate = (includeDatePicker.Pick() == 1);

	return { fileName, title, body, includeDate };
}